core/src/syscall/precompiles/blake2b/compress/mix.rs

```diff
@@ -0,0 +1,281 @@
+use crate::operations::{AddOperationU64, FixedRotateRightOperationU64, XorOperation};
+use crate::runtime::ExecutionRecord;
+
+use super::mix;
+use super::{MIX_INPUT_SIZE, STATE_ELE_PER_CALL};
+/// A set of columns needed to compute the `mix` of the input state.
+///  ``` ignore
+/// fn mix(state: &mut BlockWords, a: u64, b: u64, c: u64, d: u64, x: u64, y: u64) {
+///     state[a] = state[a].wrapping_add(state[b]).wrapping_add(x);
+///     state[d] = (state[d] ^ state[a]).rotate_right(32);
+///     state[c] = state[c].wrapping_add(state[d]);
+///     state[b] = (state[b] ^ state[c]).rotate_right(24);
+///     state[a] = state[a].wrapping_add(state[b]).wrapping_add(y);
+///     state[d] = (state[d] ^ state[a]).rotate_right(16);
+///     state[c] = state[c].wrapping_add(state[d]);
+///     state[b] = (state[b] ^ state[c]).rotate_right(63);
+/// }
+///  ```
+#[derive(AlignedBorrow, Default, Debug, Clone, Copy)]
+#[repr(C)]
+pub struct MixOperation<T> {
+    pub a_plus_b: AddOperationU64<T>,
+    pub a_plus_b_plus_x: AddOperationU64<T>,
+    pub d_xor_a_lo: XorOperation<T>,
+    pub d_xor_a_hi: XorOperation<T>,
+    // Rotate right by 32 bits by just shifting bytes.
+    pub c_plus_d: AddOperationU64<T>,
+    pub b_xor_c_lo: XorOperation<T>,
+    pub b_xor_c_hi: XorOperation<T>,
+    // Rotate roght by 24 bits by just shifting bytes.
+    pub a_plus_b_2: AddOperationU64<T>,
+    pub a_plus_b_2_add_y: AddOperationU64<T>,
+    pub d_xor_a_2_lo: XorOperation<T>,
+    pub d_xor_a_2_hi: XorOperation<T>,
+    // Rotate right by 16 bits by just shifting bytes.
+    pub c_plus_d_2: AddOperationU64<T>,
+    pub b_xor_c_2_lo: XorOperation<T>,
+    pub b_xor_c_2_hi: XorOperation<T>,
+    pub b_xor_c_2_rotate_right_63: FixedRotateRightOperationU64<T>,
+    /// `state[a]`, `state[b]`, `state[c]`, `state[d]` after all the steps.
+    pub result: [Word<T>; STATE_ELE_PER_CALL * 2],
+}
+
+impl<F: Field> MixOperation<F> {
+    pub fn populate(&mut self, record: &mut ExecutionRecord, input: [u32; 12]) -> [u32; 8] {
+        let mut a_lo = input[0];
+        let mut a_hi = input[1];
+        let mut b_lo = input[2];
+        let mut b_hi = input[3];
+        let mut c_lo = input[4];
+        let mut c_hi = input[5];
+        let mut d_lo = input[6];
+        let mut d_hi = input[7];
+        let x_lo = input[8];
+        let x_hi = input[9];
+        let y_lo = input[10];
+        let y_hi = input[11];
+
+        // the input in u64 later used to compare with the result of the mix function.
+        let input_u64 = [
+            (a_hi as u64) << 32 | a_lo as u64,
+            (b_hi as u64) << 32 | b_lo as u64,
+            (c_hi as u64) << 32 | c_lo as u64,
+            (d_hi as u64) << 32 | d_lo as u64,
+            (x_hi as u64) << 32 | x_lo as u64,
+            (y_hi as u64) << 32 | y_lo as u64,
+        ];
+
+        // First 4 steps.
+        {
+            // a = a + b + x.
+            (a_lo, a_hi) = self.a_plus_b.populate(record, a_lo, a_hi, b_lo, b_hi);
```

Comment:
yes, here your life would be a lot easier if all `populate` took in `u64` and also returned `u64`